### ISSUE TITLE
cpptools-srv SIGSEGV, IntelliSense engine is not responding. Using the Tag Parser instead

### ISSUE THREAD

**Issue by ghost** [microsoft/vscode-cpptools#9337]
Bug type: cpptools

**Describe the bug**
- OS and Version: Debian 11 (same issue also in Ubuntu 20.04 and 22.04)
- VS Code Version: 1.67.2
- C/C++ Extension Version: ms-vscode.cpptools Version 1.10.2 Version 1.10.2 (prerelease) 13/05/2022, 06:54:23
  The issue persists in any older cpptools version, too.
- Other extensions you installed (and if the issue persists after disabling them): none
- If using SSH remote, specify OS of remote machine:  no
- A clear and concise description of what the bug is:

**2..5 SIGSEGV coredumps in cpptools-srv every time I use "go to definition (F12)".** The go-to-definition takes a long time (~5 seconds), but eventually succeeds.

journalctl -b:

```
May 20 14:06:31 host kernel: cpptools-srv[365517]: segfault at 28 ip 00000000008314b6 sp 00007fd254748190 error 6 in cpptools-srv[400000+9d9000]
May 20 14:06:31 host kernel: Code: 75 26 48 8b 40 48 eb 22 31 c9 49 89 4c 24 40 45 89 ee 4c 8b 6c 24 08 8b 6c 24 04 85 ed 0f 85 22 ff ff ff e9 22 ff ff ff 31 c0 <4c> 89 60 28 41 80 4c 24 3202 49 89 44 24 50 41 f6 c6 10 74 55 49
May 20 14:06:31 host systemd[1]: Started Process Core Dump (PID 365521/UID 0).
May 20 14:06:32 host systemd-coredump[365522]: [🡕] Process 365494 (cpptools-srv) of user 1000 dumped core.

Stack trace of thread 365517:
#0  0x00000000008314b6 n/a (/home/user/.vscode/extensions/ms-vscode.cpptools-1.10.2-linux-x64/bin/cpptools-srv + 0x4314b6)
May 20 14:06:32 host systemd[1]: systemd-coredump@124-365521-0.service: Succeeded.
May 20 14:06:32 host systemd[1]: systemd-coredump@124-365521-0.service: Consumed 1.479s CPU time.
May 20 14:06:52 host kernel: cpptools-srv[365580]: segfault at 28 ip 00000000008314b6 sp 00007f96887ee190 error 6 in cpptools-srv[400000+9d9000]
May 20 14:06:52 host kernel: Code: 75 26 48 8b 40 48 eb 22 31 c9 49 89 4c 24 40 45 89 ee 4c 8b 6c 24 08 8b 6c 24 04 85 ed 0f 85 22 ff ff ff e9 22 ff ff ff 31 c0 <4c> 89 60 28 41 80 4c 24 3202 49 89 44 24 50 41 f6 c6 10 74 55 49
May 20 14:06:52 host systemd[1]: Started Process Core Dump (PID 365581/UID 0).
May 20 14:06:53 host systemd-coredump[365582]: [🡕] Process 365565 (cpptools-srv) of user 1000 dumped core.

Stack trace of thread 365580:
#0  0x00000000008314b6 n/a (/home/user/.vscode/extensions/ms-vscode.cpptools-1.10.2-linux-x64/bin/cpptools-srv + 0x4314b6)
May 20 14:06:53 host systemd[1]: systemd-coredump@125-365581-0.service: Succeeded.
May 20 14:06:53 host systemd[1]: systemd-coredump@125-365581-0.service: Consumed 1.477s CPU time.
May 20 14:06:55 host kernel: cpptools-srv[365609]: segfault at 28 ip 00000000008314b6 sp 00007f5c4c321190 error 6 in cpptools-srv[400000+9d9000]
May 20 14:06:55 host kernel: Code: 75 26 48 8b 40 48 eb 22 31 c9 49 89 4c 24 40 45 89 ee 4c 8b 6c 24 08 8b 6c 24 04 85 ed 0f 85 22 ff ff ff e9 22 ff ff ff 31 c0 <4c> 89 60 28 41 80 4c 24 3202 49 89 44 24 50 41 f6 c6 10 74 55 49
May 20 14:06:55 host systemd[1]: Started Process Core Dump (PID 365611/UID 0).
May 20 14:06:56 host systemd-coredump[365612]: [🡕] Process 365594 (cpptools-srv) of user 1000 dumped core.

Stack trace of thread 365609:
#0  0x00000000008314b6 n/a (/home/user/.vscode/extensions/ms-vscode.cpptools-1.10.2-linux-x64/bin/cpptools-srv + 0x4314b6)
May 20 14:06:56 host systemd[1]: systemd-coredump@126-365611-0.service: Succeeded.
May 20 14:06:56 host systemd[1]: systemd-coredump@126-365611-0.service: Consumed 1.470s CPU time.
```

Stacktrace:

```
$ coredumpctl debug
PID: 365594 (cpptools-srv)
UID: 1000 (user)
GID: 1000 (user)
Signal: 11 (SEGV)
Timestamp: Fri 2022-05-20 14:06:55 CEST (2min 20s ago)
Command Line: /home/user/.vscode/extensions/ms-vscode.cpptools-1.10.2-linux-x64/bin/cpptools-srv 315151 {5FE30336-D6FE-4DB3-988A-F9BB3785F665}
Executable: /home/user/.vscode/extensions/ms-vscode.cpptools-1.10.2-linux-x64/bin/cpptools-srv
Control Group: /user.slice/user-1000.slice/user@1000.service/app.slice/snap.code.code.324b8ff1-c0c0-4f6f-9ea9-c818981dec27.scope
Unit: user@1000.service
User Unit: snap.code.code.324b8ff1-c0c0-4f6f-9ea9-c818981dec27.scope
Slice: user-1000.slice
Owner UID: 1000 (user)
Boot ID: 5f69bae2140e48f29779d226352dad87
Machine ID: d2e72380ea554287ae57ac2e95fa6e76
Hostname: host
Storage: /var/lib/systemd/coredump/core.cpptools-srv.1000.5f69bae2140e48f29779d226352dad87.365594.1653048415000000.zst
Message: Process 365594 (cpptools-srv) of user 1000 dumped core.

Stack trace of thread 365609:
#0  0x00000000008314b6 n/a (/home/user/.vscode/extensions/ms-vscode.cpptools-1.10.2-linux-x64/bin/cpptools-srv + 0x4314b6)

GNU gdb (Debian 10.1-1.7) 10.1.90.20210103-git
Copyright (C) 2021 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /home/user/.vscode/extensions/ms-vscode.cpptools-1.10.2-linux-x64/bin/cpptools-srv...

warning: Can't open file /tmp/edg0_365594 (deleted) during file-backed mapping note processing
[New LWP 365609]
[New LWP 365595]
[New LWP 365601]
[New LWP 365604]
[New LWP 365598]
[New LWP 365606]
[New LWP 365594]
[New LWP 365605]
[New LWP 365607]
[New LWP 365596]
[New LWP 365599]
[New LWP 365600]
[New LWP 365602]
[New LWP 365608]
[New LWP 365597]
[New LWP 365603]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/home/user/.vscode/extensions/ms-vscode.cpptools-1.10.2-linux-x64/bin/cpptools-sr'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00000000008314b6 in i_copy_dynamic_init(a_dynamic_init*, int, a_tree_copy_control_block*) ()
[Current thread is 1 (Thread 0x7f5c4c325640 (LWP 365609))]
(gdb) bt
#0  0x00000000008314b6 in i_copy_dynamic_init(a_dynamic_init*, int, a_tree_copy_control_block*) ()
#1  0x0000000000825e44 in i_copy_constant_full(a_constant*, a_constant*, int, a_tree_copy_control_block*) ()
#2  0x0000000000825d86 in i_copy_constant_full(a_constant*, a_constant*, int, a_tree_copy_control_block*) ()
#3  0x00000000008311f7 in i_copy_dynamic_init(a_dynamic_init*, int, a_tree_copy_control_block*) ()
#4  0x000000000083102f in copy_dynamic_init(a_dynamic_init*, int) ()
#5  0x00000000007529fd in aggr_init_constant_from_field_initializer(a_field*, a_dynamic_init*, a_type*, an_init_state*, a_source_position*) ()
#6  0x000000000076268e in aggr_init_class_remainder_if_needed(a_constant*, a_type*, a_field*, a_base_class*, an_init_state*, a_source_position*, a_field*) ()
#7  0x0000000000755d57 in aggr_init_class(an_init_component**, a_type*, an_init_state*, a_source_position*, a_constant**) ()
#8  0x00000000007602c5 in aggr_init_element_full(an_init_component**, a_type*, a_field*, an_init_state*, a_source_position*, a_constant**) ()
#9  0x0000000000761e5b in aggr_init_field(an_init_component**, a_field**, an_init_state*, a_constant*, a_source_position*) ()
#10 0x0000000000755315 in aggr_init_class(an_init_component**, a_type*, an_init_state*, a_source_position*, a_constant**) ()
#11 0x00000000007536d9 in prep_aggr_initializer(an_init_component*, a_type**, an_init_state*, an_arg_match_summary*, int) ()
#12 0x000000000092868c in prep_list_initializer(an_init_component*, a_type*, int, int, int, int, int, int, int, an_operand*, an_init_state*, an_arg_match_summary*) ()
#13 0x000000000093487e in make_initializer_list_object(an_init_component*, a_type*, a_type*, int, int, a_dynamic_init**, an_operand*, an_arg_match_summary*) ()
#14 0x00000000009296fa in prep_list_initializer(an_init_component*, a_type*, int, int, int, int, int, int, int, an_operand*, an_init_state*, an_arg_match_summary*) ()
#15 0x0000000000935d91 in prep_argument(an_init_component*, a_param_type*, a_conv_descr*, an_error_code, an_operand*) ()
#16 0x0000000000921af9 in node_for_arg_of_overloaded_function_call(an_init_component*, an_arg_match_summary*, a_param_type*, a_routine*) ()
#17 0x00000000009218f7 in adjust_overloaded_function_call_arguments(a_symbol*, int, a_type*, int, an_operand*, an_init_component*, an_arg_match_summary*, an_expr_node**) ()
#18 0x00000000007885ca in scan_ctor_arguments(a_symbol*, a_source_position*, a_type*, a_type*, int, int, int, int, a_rescan_control_block*, int, an_init_component*, an_init_component*, int*, int*, int*, int*, int*, an_operand*, a_dynamic_init**, an_expr_node**, a_source_position*) ()
#19 0x0000000000929cfa in prep_list_initializer(an_init_component*, a_type*, int, int, int, int, int, int, int, an_operand*, an_init_state*, an_arg_match_summary*) ()
#20 0x000000000079cd65 in convert_initializer(an_init_component*, a_type*, int, int, an_init_state*) ()
#21 0x00000000007565a3 in braced_initializer(a_type*, an_init_component*, an_init_state*, a_decl_parse_state*, int, an_init_component**, a_source_position*) ()
#22 0x0000000000758f94 in brace_init_variable(a_decl_parse_state*, int, an_id_linkage_kind, a_source_position*, a_decl_pos_block*) ()
#23 0x0000000000757863 in initializer(a_decl_parse_state*, a_source_position*, an_id_linkage_kind, int, int*, a_decl_pos_block*) ()
#24 0x000000000074e3e4 in scan_nonmember_declaration(a_decl_parse_state*, a_source_range*) ()
#25 0x000000000075177a in namespace_declaration(a_token_kind_tag*, int, int, a_source_position*, a_symbol**) ()
#26 0x0000000000748f08 in scan_nonmember_declaration(a_decl_parse_state*, a_source_range*) ()
#27 0x000000000075177a in namespace_declaration(a_token_kind_tag*, int, int, a_source_position*, a_symbol**) ()
#28 0x0000000000748f08 in scan_nonmember_declaration(a_decl_parse_state*, a_source_range*) ()
#29 0x000000000075177a in namespace_declaration(a_token_kind_tag*, int, int, a_source_position*, a_symbol**) ()
#30 0x0000000000748f08 in scan_nonmember_declaration(a_decl_parse_state*, a_source_range*) ()
#31 0x000000000074f4c8 in translation_unit() ()
#32 0x00000000009c9e54 in process_translation_unit(char const*, int, an_exported_template_file*) ()
#33 0x00000000006a7e67 in cfe_main(int, char**) ()
#34 0x00000000009ee446 in cfe_main_exception_handler(int, char**) ()
#35 0x00000000006a7f76 in edg_main(int, char**) ()
#36 0x00000000009e77dc in edge_compiler_main(int, char const**) ()
#37 0x0000000000633656 in preparse(int, char const**, a_scout_store*, edge::translation_unit*) ()
#38 0x00000000005f82ef in a_compiler_thread::compiler_thread_routine(a_compiler_thread*) ()
#39 0x0000000000aed26a in msvc::thread_helper_t::thread_entry(void*) ()
#40 0x00007f5c50d44d80 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#41 0x00007f5c50c5e76f in clone () from /lib/x86_64-linux-gnu/libc.so.6
(gdb)
```

**Steps to reproduce**

Unfortunately the project is huge and private, can't share it.

`C_cpp_properties.json`:

```
{
    "configurations": [
        {
            "name": "mcu",
            "intelliSenseMode": "linux-gcc-x64",
            "compilerPath": "/usr/bin/gcc",
            "cStandard": "gnu17",
            "cppStandard": "gnu++17",
            "defines": [
                "TARGET_HOST",
                "STM32G474xx",
                "CPU_MIMXRT1064CVJ5A",
                "CPP_SD_EVENT",
                "CPP_SD_JOURNAL"
            ],
            "includePath": [
                "${workspaceFolder}",
                "${workspaceFolder}/.HOST-BUILD",
                "${workspaceFolder}/pm/host/.HOST-BUILD",
                "${workspaceFolder}/aux/host/mgr",
                "${workspaceFolder}/aux/host/svc",
                "${workspaceFolder}/aux/host/test",
                "${workspaceFolder}/aux/inc",
                "${workspaceFolder}/auxps/host/mgr",
                "${workspaceFolder}/auxps/host/svc",
                "${workspaceFolder}/auxps/host/test",
                "${workspaceFolder}/auxps/inc",
                "${workspaceFolder}/chassis/src",
                "${workspaceFolder}/chassis/test",
                "${workspaceFolder}/lib/canopen/src",
                "${workspaceFolder}/lib/coaxx/src",
                "${workspaceFolder}/lib/common/inc",
                "${workspaceFolder}/lib/imxrt",
                "${workspaceFolder}/lib/imxrt/drivers",
                "${workspaceFolder}/lib/imxrt/freertos-port",
                "${workspaceFolder}/lib/imxrt/lwip-port",
                "${workspaceFolder}/lib/imxrt/xip",
                "${workspaceFolder}/lib/lwip/src/include",
                "${workspaceFolder}/lib/stm/CMSIS/Device/ST/STM32G4xx/Include",
                "${workspaceFolder}/lib/stm/CMSIS/Include",
                "${workspaceFolder}/lib/stm/STM32G4xx_HAL_Driver/Inc",
                "${workspaceFolder}/lib/stm/sys",
                "${workspaceFolder}/lib/testclient/src",
                "${default}"
            ]
        }
    ],
    "version": 4
}
```

- Logs from running `C/C++: Log Diagnostics` from the VS Code command palette

Nothing special here.

```
-------- Diagnostics - 5/20/2022, 2:51:36 PM
Version: 1.10.2
Current Configuration:
<...skip...>(same as C_cpp_properties.json above)
No active translation units.

------- Workspace parsing diagnostics -------
Number of files discovered (not excluded): 40322
```

- Logs from [the language server logging](https://code.visualstudio.com/docs/cpp/enable-logging-cpp#_enable-logging-for-the-language-server)

```
IntelliSense engine is not responding. Using the Tag Parser instead.
using Tag Parser for quick info
IntelliSense process crash detected.
IntelliSense engine is not responding. Using the Tag Parser instead.
IntelliSense process crash detected.
Checking for syntax errors: /home/user/mcu/chassis/src/shell.cpp
Update IntelliSense time (sec): 0
```

Tried switching `intelliSenseMode` to `linux-clang-x64`, as well as `compilerPath` to gcc-9, gcc-10, gcc-11, clang-10, clang-11 - no difference whatsoever.

**Comment by sean-mcmanus** [comment]
We have several IntelliSense bug fixes pending for our 1.10.3 release, next Monday or Tuesday, including crash fixes: https://github.com/Microsoft/vscode-cpptools/issues?q=is%3Aopen+is%3Aissue+milestone%3A1.10.3+label%3A%22Visual+Studio%22 . So it may be fixed with that release. As a workaround, you may want to downgrade to the non-pre-release 1.9.8 until then.

**Comment by sean-mcmanus** [comment]
If it's not fixed in our pending 1.10.3, we'll look into if we can figure out the repro/cause without a sample repro code or not.
